File: arrays/05_depth_first_search/depth_first_search.py
import numpy as np


class Solution:
    def _depth_first_search(self, i: int, j: int) -> int:
        if self.path[i, j] > 0:
            return self.path[i, j]

        steps = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        for dx, dy in steps:
            x, y = i + dx, j + dy
            if (0 <= x < self.m) and (0 <= y < self.n) and (self.mat[x, y] > self.mat[i, j]):
                self.path[i, j] = max(self.path[i, j], self._depth_first_search(x, y))

        self.path[i, j] += 1
        return self.path[i, j]

    def longest_increasing_path(self, matrix: np.ndarray) -> int:
        if not self._check_constraints(matrix):
            raise ValueError('Constraints violated')

        self.path = np.zeros_like(matrix, dtype=np.uint16)
        self.mat = matrix
        self.m, self.n = matrix.shape
        max_length = 1
        for i in range(self.m):
            for j in range(self.n):
                max_length = max(max_length, self._depth_first_search(i, j))

        return max_length

    def _check_constraints(self, mat: np.ndarray) -> bool:
        if mat.shape[0] == 0 or mat.shape[0] > 200:
            return False

        for row in mat:
            if row.shape[0] == 0 or row.shape[0] > 200:
                return False

            for num in row:
                if num < 0 or num > 2**32 - 1:
                    return False

        return True

File: arrays/05_depth_first_search/depth_first_search.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <vector>

using namespace std;


class Solution {
public:
    int longestIncreasingPath(const vector<vector<int>>& matrix) {
        if (!checkConstraints(matrix)) {
            throw std::runtime_error("Constraints violated");
        }
        m = matrix.size();
        n = matrix[0].size();

        path.resize(m);
        for (auto& row : path) {
            row.assign(n, 0);
        }
        int max_path = 1;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                max_path = max(max_path, depth_search_first(i, j, matrix));
            }
        }
        return max_path;
    }

private:
    vector<vector<int>> path;
    int m, n;

    int depth_search_first(int i, int j, const vector<vector<int>>& mat) {
        if (path[i][j] > 0) return path[i][j];
        vector<pair<int, int>> steps = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        for (const auto& step : steps) {
            int x = i + step.first;
            int y = j + step.second;
            if ((x >= 0 && x < m) && (y >= 0 && y < n) && (mat[x][y] > mat[i][j])) {
                path[i][j] = max(path[i][j], depth_search_first(x, y, mat));
            }
        }
        return ++path[i][j];
    }

    // Helper function to check the constraints
    bool checkConstraints(const vector<vector<int>>& mat) {
        if (mat.size() == 0 || mat.size() > 200) return false;

        for (const auto& row : mat) {
            if (row.size() == 0 || row.size() > 200) return false;

            for (int num : row) {
                if (num < 0 || num > pow(2, 32) - 1) return false;
            }
        }
        return true;
    }
};
